Add additional tests for Matter update entity

agners opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Extend test coverage for Matter update entity. This includes tests for error handling and state store/restore.



## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

agners committed:
Add additional tests for Matter update entity

Extend test coverage for Matter update entity. This includes tests for
error handling and state store/restore.

jvmahon commented:
I did a quick test of the update feature by loading the Matter integration as a custom integration and found an issue that needs some attention:

**Issue:**
IT appears that the code does not check the "SoftwareVersionValid" field described in Section 11.22.6.7 of the Matter Core Specification and allows updates using OTA files that the manufacturer has indicated are withdrawn.

**Result:**
This allowed me to update an Aqara P2 sensor with Version 1.0.1.1 of the Aqara firmware, which bricked the P2 sensor (I'm not concerned about that - these sensors are not expensive but wanted to let you know so this is fixed before general release).

**DCL Record**
Here's the DCL record for the P2 sensor.  Note that this update to 1.0.1.1 does not show up in Apple iOS Home (I thought maybe there was an iOS Home bug, but it seems the issue is the checking of SoftwareVersionValid.

![image](https://github.com/user-attachments/assets/83e3a440-f391-4503-9aad-58e783fc712f)

MartinHjelmare commented on tests/components/matter/test_update.py:
```diff
@@ -169,3 +179,131 @@ async def test_update_install(
+        )
+
+
+async def test_update_state_restore(
```
We have `mock_restore_state_with_extra_data` to help test loading of restored state.

Example:
https://github.com/home-assistant/core/blob/7348a1fd0c9c11c3ca1a75f0c1f2e186609ec2de/tests/components/zwave_js/test_update.py#L706-L739

agners replied:
I took inspiration from `test_restore_sensor_save_state` for this test.

From what I can tell `mock_restore_cache_with_extra_data` would only allow to test the restore side, but not really the saving side (making sure the update information makes it into the extra data).

The way I wrote the test is really what I (the user) is interested in: Make sure the update information survives a restart.

MartinHjelmare replied:
I didn't mean that we should remove this test. But we need to add another test that uses that helper and tests the loading.

agners replied:
I see, yeah seems sensible to test the loading side only, to make sure this remains compatible/correct.

We don't use much of the extra data currently, essentially we just use it to get the software update integer version to pass it to the update call. I guess it could be debated if the implementation of the Matter update entity should make use more of the extra data directly :thinking:  But that would need an update entity change too.

Anyways, the current test really tests what extra data is used now, so IMHO this is already a good step forward in test coverage.

MartinHjelmare commented on tests/components/matter/test_update.py:
```diff
@@ -169,3 +179,131 @@ async def test_update_install(
+    update_node: AsyncMock,
+    updateable_node: MatterNode,
+) -> None:
+    """Test update entity exists and update check got made."""
```
This test is testing state restore saving, right?

MartinHjelmare commented on tests/components/matter/test_update.py:
```diff
@@ -169,3 +179,131 @@ async def test_update_install(
+    update_node: AsyncMock,
+    updateable_node: MatterNode,
+) -> None:
+    """Test update entity exists and update check got made."""
```
```suggestion
    """Test update entity service call errors."""
```

MartinHjelmare commented:
@jvmahon please open an issue instead.

MartinHjelmare commented:
Looks like two errors and the polling mechanism are still uncovered.

marcelveldt commented:
@jvmahon like Martin already noted: don't start commenting on random/unrelated PR's and/or issue reports, create your own issue report if you find something. I know you are trying to be helpful in these cases but please respect our developer workflow. 

To quickly answer your message about that P2 sensor. We are in fact checking (server side) for "software version valid" but Aqara did not set that. They just did on our request so probably your device is now bricked because you updated it right before that. NOTE: Maybe its not bricked and you just need to to reset and recommission it.

agners committed:
Improve test descriptions

agners committed:
Add restore test only (using mock_restore_cache_with_extra_data)

agners committed:
Fix test_update_state_save_and_restore test

joostlek commented on tests/components/matter/test_update.py:
```diff
@@ -169,3 +194,185 @@ async def test_update_install(
+    assert state.state == STATE_OFF
+    assert state.attributes.get("installed_version") == "v1.0"
+
+    await async_setup_component(hass, "homeassistant", {})
```
```suggestion
    await async_setup_component(hass, HA_DOMAIN, {})
```

joostlek commented on tests/components/matter/test_update.py:
```diff
@@ -169,3 +194,185 @@ async def test_update_install(
+        update_source=UpdateSource.LOCAL,
+    )
+
+    await hass.services.async_call(
+        "homeassistant",
+        "update_entity",
+        {
+            ATTR_ENTITY_ID: "update.mock_dimmable_light",
+        },
+        blocking=True,
+    )
```
How does this normally get triggered? I'd rather see it naturally be updated using either getting a push function from a mock and calling it or skipping time

agners replied:
It is polled normally, every 12h:

https://github.com/home-assistant/core/blob/98a007cb2f04eaba6bc28651d964816f45e57724/homeassistant/components/matter/update.py#L73-L76

Using the `update_entity` action is a perfectly valid way to force an update IMHO. But if there is a easy way to test that poll, that is probably closer to the typical use case. :thinking:

joostlek replied:
We can use the freezer to skip 12 hours in time and then it will automatically update

agners replied:
Sounds good! I've added a separate test to test the service call still, and use the freezer by default.

joostlek commented on tests/components/matter/test_update.py:
```diff
@@ -169,3 +194,185 @@ async def test_update_install(
+    )
+
+    await hass.services.async_call(
+        "homeassistant",
+        "update_entity",
```
Otherwise, please use constants
```suggestion
        HA_DOMAIN,
        SERVICE_UPDATE_ENTITY,
```

joostlek commented on tests/components/matter/test_update.py:
```diff
@@ -169,3 +194,185 @@ async def test_update_install(
+
+    with pytest.raises(HomeAssistantError):
+        await hass.services.async_call(
+            "update",
+            "install",
+            {
+                ATTR_ENTITY_ID: "update.mock_dimmable_light",
+                "version": "v3.0",
+            },
+            blocking=True,
+        )
+
+    update_node.side_effect = UpdateError("Error updating node")
+
+    with pytest.raises(HomeAssistantError):
+        await hass.services.async_call(
+            "update",
+            "install",
+            {
+                ATTR_ENTITY_ID: "update.mock_dimmable_light",
+                "version": "v3.0",
```
Also please use constants here

agners replied:
Is using constants really a good idea in all cases? E.g. when using constants in a test, a change of that constant might actually break the user facing API (e.g. in this case users might have `version` hard coded in an automation or something). So changing the constant should break this test. So a hard coded value is better, IMHO.

agners replied:
Hm, I guess that is the API of the update component, so a breaking change would not really be a concern of this particular implementation (on the Matter end).

I've replaced things with constants :+1:

joostlek commented on tests/components/matter/test_update.py:
```diff
@@ -169,3 +194,185 @@ async def test_update_install(
+    assert state.attributes.get("latest_version") == "v2.0"
+
+    await hass.services.async_call(
+        "update",
+        "install",
```
idem

agners committed:
Use homeassistant constants

agners committed:
Use update component constants

agners committed:
Use freezer to skip time for device update check

We check device updates every 12h currently. Use the freezer to skip
time.

Still add a test which uses the service call to make sure this works
too.
